### DISCUSSION TITLE
Request for Enhancement: Separate SwapItemsAsync and Solve Board Functions

### DISCUSSION THREAD

**Discussion by Unshackled-Games** [ChebanovDD/MatchSweets#17] [Ideas]
Good Evening,

I want to express my gratitude for this amazing API; it's exceptionally well-done. However, I've noticed a challenge in customizing the behavior to better suit specific needs. Currently, the SwapItemsAsync function also triggers the board-solving process. This means that it's impossible to swap items without inadvertently initiating the solve function, and vice versa – solving the board triggers item swapping.

I believe it would greatly enhance the flexibility and usability of the API if these two functionalities could be separated into distinct functions. This separation would allow users to independently trigger either the item swapping or board-solving process as needed.

Thank you for considering this enhancement. I believe it would make the API even more versatile and user-friendly.

**Comment by ChebanovDD** [comment]
Hello.

Thank you for your suggestion. These two functions are combined into one to handle the scenario where the board can not be "solved" during the swap; in such cases, the items are swapped back. If we were to separate these functions, it would require manual monitoring for this specific condition.

I'm open to suggestions on how to make the API more user-friendly. Do you have any idea how this can be implemented?

**Comment by Unshackled-Games** [comment]
Hello 

Thank you for your response. Following the design the swap will be called through the Input system but as the API is at the moment you can only swap two specific positions and trigger the solve. Would it make sense to add an layer to the Input system? 

At the moment the 
``` csharp
public interface IInputSystem
    {
        event EventHandler<PointerEventArgs> PointerDown;
        event EventHandler<PointerEventArgs> PointerDrag;
        event EventHandler<PointerEventArgs> PointerUp;
    }
```

But would it make sense to have one Interface that handles the trigger for these two behaviour's swap and solve? 
Like when you want to swap and solve you would trigger swap and after that solve in the PointerDrag but then when you want to have a different behaviour you could Implement your own class with this new Interface to call the solve only when PointerUp or implement your own conditions? 

I am sorry that my response is so vague. I don't think I have a real solution for the difficult case but I wanted to provide some thoughts on that matter. What do you think?

**Comment by ChebanovDD** [comment]
The `SwapItemsAsync` method is now virtual. You can override it and write any logic for item-swapping 🙂

**Reply by Unshackled-Games** [reply]
Thank you :). That helps a lot!
